Add WH-1000XM5, WH-1000XM6, and WF-C510 support

Amrsatrio opened the pull request:
This PR adds support to those models, mainly for WH-1000XM6:
* Battery indicator
* NC/AMB get & notify support (+ auto ambient level)
  * Revamped this page to better match the app.
* Listening Mode
* Touch sensor control panel
* Pause when taken off
* NC/AMB button configuration

TODO:
* ~~Fix the rest such as equalizer.~~
* ~~Determine a better method of detecting the headphones' model. I added a simple device detection based on the data in INIT_RESPONSE and enable certain code paths based on that. But since I only own a WF-1000XM5 with me, I can't guarantee compatibility with other models. Might want to give a shot in reversing the iOS/Android app.~~
* Test on WF-CH720N and ~~WF-C510~~.

Bonus:
* Fixed Sound Pressure outright not working.
* Fixed voice guidance -2 showing as 254.
* Introduced a payload object serialization system with least heap allocations as possible.
  * Payloads using this system are now validated.
  * Those payload objects were derived from the iOS version of the Sound Connect app.

<img width="904" height="741" alt="image" src="https://github.com/user-attachments/assets/878c5f89-11fb-4013-922f-af47688590a0" />
<img width="904" height="741" alt="image" src="https://github.com/user-attachments/assets/2bec694a-436c-42a6-b518-8ab752b76d6e" />
<img width="904" height="741" alt="image" src="https://github.com/user-attachments/assets/6b685d98-7514-41c5-b163-170a2eda0957" />
<img width="904" height="741" alt="image" src="https://github.com/user-attachments/assets/34dcb52b-037b-4532-aacf-56d5a9e3eed1" />

### WH-100XM6 feature table

| Feature                                  | Status                              |
|------------------------------------------|-------------------------------------|
| Current Playing                          | ✅                                   |
| Sound Pressure                           | ✅                                   |
| Battery Life                             | ✅                                   |
| Volume                                   | ✅                                   |
| NC/AMB Settings                          | ✅ (supports new auto ambient level) |
| Voice Guidance Volume                    | ✅                                   |
| Track Controls                           | ✅                                   |
| Multipoint Control                       | ~                                   |
| Speak to Chat                            | **?** (not tested)                  |
| Listening Mode                           | ✅                                   |
| Equalizer                                | ✅                                   |
| Touch Sensor Enable/Disable              | ✅                                   |
| NC/AMB Button Setting                    | ✅                                   |
| Power Off                                | ✅                                   |
| Capture Voice During a Phone Call        | ✅                                   |
| Map Headphone Gestures to Shell Commands | **?** (not tested)                  |

mos9527 commented on src/Headphones.h:
```diff
@@ -84,7 +86,18 @@ enum class HeadphonesEvent
+
 using HeadphonesMessage = CommandSerializer::CommandMessage;
+
+enum DeviceCapabilities
```
Maybe you meant enum class? 🤔 C enums don't need bitwise overloads

Amrsatrio replied:
In C they don't, but in C++ they do. This is what Microsoft does to enums in their codebase too.
And enum classes do not implicitly cast to their underlying type making me have to do something like this to test if a flag is active or not:
```cpp
(deviceCapabilities & DeviceCapabilities::DC_AutoAsm) != DeviceCapabilities::DC_None
```
Can't do this:
```cpp
(deviceCapabilities & DeviceCapabilities::DC_AutoAsm) != 0
```

(At least with the bitwise overloads I adapted from the Windows SDK)

Amrsatrio commented:
If Android is obfuscated *that* badly, I have access to a TrollStore-compatible iOS device so I dump a decrypted IPA and try to see how it works. The problem is there are too many dylib's and pretty hard to find which is which...

Amrsatrio commented:
<img width="904" height="838" alt="image" src="https://github.com/user-attachments/assets/d669fcd6-64df-4432-9a54-f99d2d749a21" />
Fixed EQ.

Amrsatrio commented:
<img width="904" height="633" alt="image" src="https://github.com/user-attachments/assets/fc4a81e1-81b5-4855-87b6-b19de1580bbd" />
Added some more options that I tested to work on the WF XM5 as well.

mos9527 commented:
> Cracked open the iOS app and found that it's still written in Java transpiled into Objective-C using j2objc. All the names are there too. And I got the full command names -> ids for V1 and V2. Do you think it's fine to switch to the official names?

@Amrsatrio That's some great news! And yes, I think it should be fine to use the ones from offical tables considering we're only starting from the offical binaries. I've found some other MDR implementations on GitHub that uses these names as well:
- https://github.com/AndreasOlofsson/libmdr
- https://github.com/Keller18306/SonyHeadphonesProtocol

Amrsatrio commented:
TODO:
- Fix voice capture not hidden on WH-1000XM5
- Fix touch sensor control panel toggle not shown on WH-1000XM5 (instead controls for WF are shown)

Amrsatrio commented:
If you want to work on the formatting I'd suggest to work on my fork now, while I figure out those two TODOs.

My suggestion for the formatting:

```cpp
    uint8_t _sumChecksum(const uint8_t* src, size_t size)
    {
        uint8_t accumulator = 0;
        for (size_t i = 0; i < size; i++)
        {
            accumulator += src[i];
        }
        return accumulator;
    }
```

* Stick with 4 spaces (like in App.cpp), using tab to add indentation may conflict with AI autocompletion tools.
* Braces in new line, so that we can distinguish if conditions easily from its block.

mos9527 commented:
> * Stick with 4 spaces (like in App.cpp), using tab to add indentation may conflict with AI autocompletion tools. Microsoft also uses 4 spaces in their C++ codebase.
> * Braces in new line, so that we can distinguish if conditions easily from its block.
> * const T&
> * Operators at new line for visibility.
> 
> *_Or if you allow me to do so, I'll do it happily :)_

@Amrsatrio Of course. Feel free to introduce formating (maybe with `.clang-format` and co?) to the codebase - what you've suggested looks good to me!

mos9527 commented on src/Headphones.h:
```diff
@@ -206,23 +331,114 @@ class Headphones
+template <typename TPayload, typename... TArgs>
+void Headphones::sendMdrCommandWithType(CommandType ct, TArgs&&... args)
+{
+	if constexpr (TPayload::VARIABLE_SIZE) {
```
I see there's `StringData`, `ConnectRetSupportFunction` payloads that's of variable sizes.  (curious - do we ever _send_ such payloads to the headphones?)  
Considering we'd allocate on the heap anyways, the `unique_ptr` indirection is uncessary - I suggest implementing dedicated serialization-to-bytes fucntions for these as there's no real benefit to _always_ alias the underlying data.

mos9527 commented on CMakeLists.txt:
```diff
@@ -103,6 +114,13 @@ if (WIN32)
+    if (MSVC)
+        # Disable RTTI
+        # Enable Multi Processor Compilation
+        target_compile_options(SonyHeadphonesClient PRIVATE /GR- /MP)
```
With `/MP` - [CMake's jobs (`-j`) option](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-build-j) should be preferred over this - with build systems like `ninja` this would be irrelevant too since they'd schedule sources one-by-one themselves, in parallel.

And with turning RTTI off - add this to other compilers too!

Amrsatrio replied:
I don't have that much experience with CMake, as well as experience with developing native code for platforms other than Windows, sorry. Might need your help for this one.

mos9527 commented on CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@ cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
 add_definitions(-D__HEADPHONES_APP_VERSION__="1.4.4")
 
-set(CMAKE_CXX_STANDARD 20)
+set(CMAKE_CXX_STANDARD 23)
```
README Building section could use a update too

Amrsatrio replied:
Noted.

Amrsatrio commented on docs/device-support/WH-1000XM5.md:
```diff
@@ -6,20 +6,18 @@ Reported by:
+| Touch Sensor Enable/Disable              | ✅                                                                                                    |
+| NC/AMB Button Setting                    | ✅                                                                                                    |
+| Power Off                                | ✅                                                                                                    |
+| Map Headphone Gestures to Shell Commands | ✅ ❌ (gui has errors preventing config update, but does work after manual TOML configuration changes) |
```
Sure.

Amrsatrio commented on src/App.cpp:
```diff
@@ -301,8 +356,60 @@ void App::_drawControls()
+                ListeningMode effectiveMode = _headphones->listeningModeConfig.current.getEffectiveMode();
+
+                // Standard
+                bool radioChanged = ImGui::RadioButton("Standard", (int*)&effectiveMode, (int)ListeningMode::Standard);
```
Alright.

Amrsatrio commented:
<img width="904" height="647" alt="image" src="https://github.com/user-attachments/assets/01b0b9cc-a8bd-41e4-ab14-bd85c3f74878" />

Added support for WF-C510 to test my feature testing code, this model only has AMB 1-20 and no NC. Implemented general settings properly, now problems with WH-1000XM5 have been ironed out.

Amrsatrio commented:
<img width="904" height="647" alt="image" src="https://github.com/user-attachments/assets/4eeb7ed3-5c41-40f2-b8bf-2dafef492bca" />

Fixed sound pressure retrieval.

Amrsatrio commented:
Ready for another review. As for this:
> Considering we'd allocate on the heap anyways, the unique_ptr indirection is uncessary - I suggest implementing dedicated serialization-to-bytes fucntions for these as there's no real benefit to always alias the underlying data.

I made it clearer that it only applies to messages that has only one variable length data which is an array at the end. For multi variable length data like GsRetCapability I also added option to use a serialization/deserialization into the new templated payload sender function I made.

mos9527 commented on src/Protocol.h:
```diff
@@ -0,0 +1,80 @@
+    buf.insert(buf.end(), str.begin(), str.end());
+}
+
+#define VARIABLE_SIZE_PAYLOAD_ONE_ARRAY_AT_END(maxSize) \
```
Macros! Though fair enough - this is clean enough to also work. 
Another idea - how about declaring a `std::array<TEndArrayElement, maxSize>` for data - considering we know that max bounds and length in the impls I've seen thus far? Or implementing seperate `serialize` `deserialize` functions with the macro instead and leverage STL types like `std::string` and `std::vector` for storage?

Amrsatrio replied:
Using `std::array<TEndArrayElement, maxSize>` expands the array variable size to `maxSize * sizeof(TEndArrayElement)`. Using the empty array declaration now does not influence the struct size but still allows data from that offset to be accessed without casting `this` to `uint8_t*`.

As for serialize/deserialize, I think it is not needed for these kind of payloads with only one variable size array. Remember, I like optimizing stuff, so I'd like stuff done with least heap allocations/reallocations as possible and less binary instructions on Release if possible.

mos9527 replied:
I see. Guess using raw `new` or not going for the same seralize pattern triggered some OCD in me. 

I'd agree with you on the usage here - this seems like a very reasonable tradeoff.

mos9527 reviewed (changes requested):
Looks good so far. Once these are done I believe I can merge this first - this PR is getting quite heavy!

Amrsatrio commented on src/Headphones.cpp:
```diff
@@ -327,114 +728,426 @@ void Headphones::recvAsync()
+{
+    if (modelName.current == "WF-C510")
+    {
+        return HeadphonesEvent::NoChange; // @FIXME Assertion error when using WF-C510
```
<img width="1079" height="1023" alt="image" src="https://github.com/user-attachments/assets/198db023-ebc3-4714-ba12-4cfaa043d77e" />

Here. I'll figure this out later including adding more operations the app's device management UI offers.

Amrsatrio commented on src/Headphones.cpp:
```diff
@@ -1,317 +1,718 @@
+    /* Init */
+    // NOTE: It's observed that playback metadata NOTIFYs (see _handleMessage) is sent by the device after this...
+    sendGet<THMSGV2T1::ConnectGetProtocolInfo>(THMSGV2T1::ConnectInquiredType::FIXED_VALUE);
+    waitForProtocolInfo(5);
```
waitAck() is now included in sendGet/Set now. Alright I'll extract the 5 into a constexpr local var.

mos9527 commented:
> Added Windows ARM64 variant. CI green.
> 
> One thing to note, the Windows builds released have inlining disabled. I guess it's better to hack CMake to build Release binaries but still generate PDBs on Windows. 

Found [this post](https://www.reddit.com/r/cpp/comments/1nh2xx3/why_does_cmake_configuration_relwithdebinfo_by/) - should be helpful to re-enable inlining for Windows release builds while still using `RelWithDebInfo` flags which CI uses by default.

> Not sure how PDBs are handled on Linux and macOS.

Linux builds have their debug info emebdded in the ELF binaries itself. CI [currently strips this](https://github.com/mos9527/SonyHeadphonesClient/blob/master/.github/workflows/cmake.yml#L82) currently as it bloats the binary by quite a bit (~10MB), though I suppose we can provide a seperate, debuggable download? On macOS it's also embedded into the executable - CI doesn't strip it as the artifacts aren't that large.

Amrsatrio commented:
<img width="904" height="379" alt="image" src="https://github.com/user-attachments/assets/b8800e89-1757-4164-bf91-3a681960484c" />

Fixed devices list retrieval. Existing code read the payload wrongly. Studied the data structures in the iOS app and it's proper now.

Amrsatrio commented:
Thanks! #4 #5 can be closed now. Don't have the device to close #7 though... must test its NC settings update when you press the NC/AMB button.

mos9527 commented:
> Thanks! #4 #5 can be closed now. Don't have the device to close #7 though... must test its NC settings update when you press the NC/AMB button.

Got it. I'll close these once I'd release 1.4.5 in a (hopefully) not-so-distant future 👍 Guess I'll leave #7 open for now until someone could test it out on the new versions.

Amrsatrio commented:
We must note out that using LE audio (my Snapdragon X Elite laptop supports it) causes the program's RFCOMM connection to the device to outright not work at all.

By the way, do you have a DIscord or such where we can communicate instantly? Would be great if we have people with other kinds of devices than I have to test our changes.

mos9527 commented:
> We must note out that using LE audio (my Snapdragon X Elite laptop supports it) causes the program's RFCOMM connection to the device to outright not work at all.

Curious - maybe we should open an Issue ticket for this. https://github.com/mos9527/SonyHeadphonesClient/issues/18 probably isn't releavant, but could suggest issues with our current Windows BT socket impl  (which I've basically kept as is from the [original](https://github.com/Plutoberth/SonyHeadphonesClient))

> By the way, do you have a DIscord or such where we can communicate instantly? Would be great if we have people with other kinds of devices than I have to test our changes.

You can find me with the same handle (mos9527) there. Though I'd imagine finding volunteers to beta test could be difficult - the tool isn't that popular after all. I suppose GitHub Issues would be sufficent for such purposes 🤔

On that note, I'll update the README to include a nightly download URL. Thanks for the reminder!

Amrsatrio commented on src/App.cpp:
```diff
@@ -286,23 +331,137 @@ void App::_drawControls()
+
+                // Standard
+                bool radioChanged = false;
+                if (radioChanged |= ImGui::RadioButton("Standard", effectiveMode == ListeningMode::Standard))
```
Just realized that `if (radioChanged |= ImGui::RadioButton(...)) ...` is wrong. Should've been `if (ImGui::RadioButton(...)) { radioChanged |= true; ... }`

I'll address this in an upcoming PR.
